src/client/dfuse/dfuse_core.c

```diff
@@ -21,9 +21,48 @@
  * portions thereof marked with this legend must also reproduce the markings.
  */
 
+#include <pthread.h>
+
 #include "dfuse_common.h"
 #include "dfuse.h"
 
+static void *
+dfuse_progress_thread(void *arg)
+{
+	struct dfuse_projection_info *fs_handle = arg;
+	int rc;
+	daos_event_t *dev;
+	struct dfuse_event *ev;
+
+	while (1) {
+
+		errno = 0;
+		rc = sem_wait(&fs_handle->dpi_sem);
```

Comment:
TBH i don't understand why you need semaphores. could you please explain a little or add some comments in the code about this.

Reply:
I'll add some more comments in-line but essentially I don't want dfuse to consume any cycles when there are no pending messages.